flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractHandler.java

```diff
@@ -114,6 +115,21 @@ protected void respondAsLeader(ChannelHandlerContext ctx, RoutedRequest routedRe
 			log.trace("Received request " + httpRequest.uri() + '.');
 		}
 
+		synchronized (this) {
+			if (terminationFuture != null) {
+				String errorMsg = "The handler instance for " + untypedResponseMessageHeaders.getTargetRestEndpointURL()
+					+ " had already been closed";
+				log.warn(errorMsg);
+				HandlerUtils.sendErrorResponse(
```

Comment:
We should not try to submit a response; since we are either already shutdown or in the process of shutting down the response may or may not go through. Better to be consistent and not respond in either case.
One thing we can do though is close the channel via `ctx.channel().close()`.

Reply:
If we doesn’t submit a response，the client only wait for timeout.
and if we do ```ctx.channel().close()``` , how about the request in flight?

Reply:
As of right now, should the NPE occur then all the user/dev sees is the same stack trace you got. That's not ideal.

So instead, let's handle the case explicitly:
```
private CompletableFuture<Void> handleException(Throwable throwable, ChannelHandlerContext ctx, HttpRequest httpRequest) {
		FlinkHttpObjectAggregator flinkHttpObjectAggregator = ctx.pipeline().get(FlinkHttpObjectAggregator.class);
		if (flinkHttpObjectAggregator == null) {
			log.warn("The connection was closed unexpectedly.");
			return CompletableFuture.completedFuture(null);
		}
		...
```

Reply:
Do we still need to modify the processing logic of ```requestProcessingFuture```? If we avoid NPE, the changes to ```requestProcessingFuture``` may not be necessary.

Reply:
they are still useful for preventing   something similar from happening again.

Reply:
yes